### DISCUSSION TITLE
Failure to process HEIC images taken on iOS18 iPhone

### DISCUSSION THREAD

**Discussion by PearlPig123** [ImageMagick/ImageMagick#7426] [Development]
### ImageMagick version

7.1.0-61

### Operating system, version and so on

Common Base Linux Mariner 2.0.20240223 #59-Ubuntu

### Description

Hi imagemagick experts, I found an issue related to HEIC images taken by iOS18 iPhone where the magick command would fail with the below error:
```
$ magick identify -format "%m,%w,%h\n%[EXIF:*]" failure_magick.HEIC

identify: Invalid input: Unspecified: Metadata not correctly assigned to image (2.0) `failure_magick.HEIC' @ error/heic.c/IsHEIFSuccess/139.
```
I have attached the link to an example failure image: https://drive.google.com/file/d/1MuoNixR8mNlq-cEd3OMSWB8K3bV_qCZm/view?usp=sharing
The issue started with iOS 18 only, the photos taken on iOS 17 can be processed without problem.
Wonder if it is possible to fix this issue? Thank you!

**Comment by fmw42** [comment]
Fails for me also on IM 7.1.1.34 with libheif 1.17.6 on Mac OSX Ventura.  But it opens in Mac Preview and also GraphicConverter.

**Comment by dlemstra** [comment]
Attaching file for future reference [failure_magick.zip](https://github.com/user-attachments/files/15982700/failure_magick.zip).

It is possible that the libheif library that we are using has no support for reading this format.

**Reply by PearlPig123** [reply]
Thanks for checking! The image was taken by an iPhone recently upgraded to iOS 18, I suspect that Apple changed the way they encode HEIC in iOS 18 but I couldn't find any of this in their release notes. 
I was using this [libheif](https://github.com/strukturag/libheif) library to build my imagemagick command line on linux. I wonder if the issue is from imagemagick or libheif? And should I convert this to an issue?

**Reply by dlemstra** [reply]
Libheif is telling us that it cannot read the file so this is a libheif issue and not an ImageMagick issue. You should ask them for support of this format.

**Reply by Gillardo** [reply]
Did you get a fix for this in the end?  We are experiencing the same issue when trying to open a HEIC file taken on a iPhone 15 Pro Mox

**Reply by dlemstra** [reply]
It seems that the latest version of libheif has support for this file:

```
>identify failure_magick.HEIC
failure_magick.HEIC HEIC 4032x3024 4032x3024+0+0 8-bit sRGB 1.08027MiB 0.034u 0:00.033
```

**Comment by whizzzkid** [comment]
Relevant: https://xkcd.com/2347/

**Comment by tobiaseigen** [comment]
I am getting a similar error on new heif images created since I upgraded my phone last week, using the Nextcloud AIO and the memories app. Not sure how to proceed - if anyone has insights I'd appreciate! 🤗

**Comment by snibgo** [comment]
> Not sure how to proceed ...

You could upload a sample input image file, the command you used, and tell us what version of IM you used, and the text output from "magick -list format |grep -i heif".

**Reply by tobiaseigen** [reply]
Thanks for the suggestion! Hope this helps you help me. I appreciate it!  

I honestly don't know what the command is that nextcloud runs to generate previews and thus is generating this error. Here are examples of the specific errors I am getting. One oddity is that the `/mnt/ncdata...` path is incorrect. 

```
ImagickException
Invalid input: Unspecified: Metadata not correctly assigned to image (2.0) `/mnt/ncdata/kb2-1/files/Tobias Photos/2024/12/24-12-03 12-49-22 7499.heic' @ error/heic.c/IsHEIFSuccess/144
File: /kb2-1/files/Tobias Photos/2024/12/24-12-03 12-49-22 7499.heic Imagick says:
```

Here's the version info: 

```
root@2024-nextcloud-ubuntu-s-2vcpu-4gb-120gb-intel-nyc3-01:~# sudo docker exec --user www-data -it nextcloud-aio-nextcloud magick --version
Version: ImageMagick 7.1.1-32 Q16-HDRI x86_64 22207 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5) 
Delegates (built-in): bzlib cairo fftw fontconfig freetype gslib heic jng jpeg jxl lcms ltdl lzma pangocairo png ps raw rsvg tiff webp x xml zlib zstd
Compiler: gcc (13.2)
```

result of the command you asked for: 

```
root@2024-nextcloud-ubuntu-s-2vcpu-4gb-120gb-intel-nyc3-01:~# sudo docker exec --user www-data -it nextcloud-aio-nextcloud magick -list format |grep -i heif
     HEIF  HEIC      rw+   High Efficiency Image Format (1.17.6)
```

[24-12-03 08-50-07 7496.heic.zip](https://github.com/user-attachments/files/18013365/24-12-03.08-50-07.7496.heic.zip)

**Comment by drgrib** [comment]
I can confirm that upgrading my `brew` installation using `upgrade` fixed the issue for me:
```
brew upgrade imagemagick
```

**Reply by ashwinvidiyala** [reply]
Worked for me too. Thank you kind stranger @drgrib

**Comment by snibgo** [comment]
> Permission denied @ error/blob.c/OpenBlob/3596

This suggests the OS is denying ImageMagick permission to access the file. The problem may be at the file or directory level. "chmod" might fix that.

> ... upgrading my brew installation using upgrade fixed the issue ...

Good stuff. So your problem is solved? 

On my Windows 11 running IM v7.1.1-38 with libheif 1.18.2, ...
```
magick "24-12-03 08-50-07 7496.heic" x.png
```
... works with no problem.

**Reply by tobiaseigen** [reply]
> > Permission denied @ error/blob.c/OpenBlob/3596

I noticed that I had pasted the wrong error - and indeed that is about permissions. I replaced it with the one that I am concerned about, which aligns with what we see here: 

```
ImagickException
Invalid input: Unspecified: Metadata not correctly assigned to image (2.0) `/mnt/ncdata/kb2-1/files/Tobias Photos/2024/12/24-12-03 12-49-22 7499.heic' @ error/heic.c/IsHEIFSuccess/144
File: /kb2-1/files/Tobias Photos/2024/12/24-12-03 12-49-22 7499.heic Imagick says:
```

**Comment by regularRandom** [comment]
Version 7.1.1.41-1 has the same issue:

> ImageMagick says: Invalid input: Unspecified: Metadata not correctly assigned to image (2.0) `\/tmp\/magick-DHPOFENZ2nZkj8pEr4b2X9bakhWrWnoE' @ error\/heic.c\/IsHEIFSuccess\/141

I have CentOS Stream release 9, ownCloud 10.15.0 (stable), ImageMagick7-heic-7.1.1.41-1.el9.remi.x86_64.

**Comment by snibgo** [comment]
From posts on this thread, it seems the problem isn't with IM, but old versions of libheif: 1.17.6 can't read the posted file, but 1.18.2 can. "magick -list format" will tell you what version of libheif is used by your IM.

**Comment by regularRandom** [comment]
@snibgo yes, exactly. I just rebuilt libheif from the source and voila - everything started working!

**Comment by RGD2** [comment]
Try this: https://strukturag.github.io/libheif/

This is a webpage embedded viewer using the latest version of the libheif library: Good enough as a workaround if you just need to convert a couple images to JPG, while you wait for the updated library version to percolate down.
